### PR TITLE
Forum: Bugfix 0033198: first posting missing when copying a forum

### PR DESCRIPTION
**Pull request by fhelfer**
This fixes https://mantis.ilias.de/view.php?id=33198

### CODE REVIEW

File: Modules/Forum/classes/class.ilForumTopic.php [hunk 1]
```diff
@@ -215,13 +215,13 @@ public function reload() : bool
     public function getFirstPostId() : int
     {
         $res = $this->db->queryF(
-            'SELECT * FROM frm_posts_tree WHERE thr_fk = %s AND parent_pos = %s',
-            ['integer', 'integer'],
-            [$this->id, 1]
+            'SELECT * FROM frm_posts_tree WHERE thr_fk = %s AND parent_pos != %s AND depth = %s ORDER BY rgt DESC',
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumTopic.php]: - Shouldn't we add an `$this->db->setLimit(1);` here as well?

File: Modules/Forum/classes/class.ilForumTopic.php [hunk 2]
```diff
@@ -285,15 +286,18 @@ public function countActivePosts(bool $ignoreRoot = false) : int
...
     {
+        $this->db->setLimit(1);
         $res = $this->db->queryF(
             '
 			SELECT *
 			FROM frm_posts 
 			INNER JOIN frm_posts_tree ON pos_fk = pos_pk
-			WHERE parent_pos = %s
-			AND thr_fk = %s',
+			WHERE parent_pos != %s
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumTopic.php:335]: IMO This is wrong in this case, I checked the usages. We should rename this method to `getPostRootNode`. The `depths` MUST be 1 and the `parent_pos` MUST equal `0`.

File: Modules/Forum/classes/class.ilForumTopic.php [hunk 3]
```diff
@@ -214,14 +214,15 @@ public function reload() : bool
 
     public function getFirstPostId() : int
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumTopic.php]: Maybe we should rename it to `getFirstVisiblePostId` (see comment below).

**Review by mjansenDatabay — CHANGES_REQUESTED** [review]

File: Modules/Forum/classes/class.ilForumTopic.php (hunk 3 above)
Comment (fhelfer) [reply Modules/Forum/classes/class.ilForumTopic.php]: we could add a optional parameter for both methods ($exclude_root_node = false)

Comment (mjansenDatabay) [reply Modules/Forum/classes/class.ilForumTopic.php]: Hmmm, I am personally not a fan of introducing parameters (we do it to often in legacy code).

> Don't use flag arguments. Split method into several independent methods that can be called from the client without the flag.

File: Modules/Forum/classes/class.ilForumTopic.php [hunk 4]
```diff
@@ -212,16 +212,32 @@ public function reload() : bool
...
 
-    public function getFirstPostId() : int
+    public function getPostRootId() : int
     {
+        $this->db->setLimit(1);
         $res = $this->db->queryF(
-            'SELECT * FROM frm_posts_tree WHERE thr_fk = %s AND parent_pos = %s',
-            ['integer', 'integer'],
-            [$this->id, 1]
+            'SELECT * FROM frm_posts_tree WHERE thr_fk = %s AND parent_pos = %s AND depth = %s ORDER BY rgt DESC',
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumTopic.php]:
```suggestion
            'SELECT pos_fk FROM frm_posts_tree WHERE thr_fk = %s AND parent_pos = %s AND depth = %s ORDER BY rgt DESC',
```

File: Modules/Forum/classes/class.ilForumPost.php [hunk 5]
```diff
@@ -113,7 +113,7 @@ public function update() : bool
                 ]
             );
 
-            if ($this->objThread->getFirstPostId() === $this->id) {
+            if ($this->objThread->getPostRootId() === $this->id) {
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumPost.php]: The first visible posting id should be used here.

**Review by mjansenDatabay — CHANGES_REQUESTED** [review]

File: Modules/Forum/classes/class.ilForumTopic.php [hunk 6]
```diff
@@ -212,16 +212,32 @@ public function reload() : bool
...
+            return (int) $row->pos_fk ?: 0;
+        }
+        return 0;
+    }
+
+    public function getFirstVisiblePostId() : int
+    {
+        $this->db->setLimit(1);
+        $res = $this->db->queryF(
+            'SELECT * FROM frm_posts_tree WHERE thr_fk = %s AND parent_pos != %s AND depth = %s ORDER BY rgt DESC',
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumTopic.php]:
```suggestion
            'SELECT pos_fk FROM frm_posts_tree WHERE thr_fk = %s AND parent_pos != %s AND depth = %s ORDER BY rgt DESC',
```

File: Modules/Forum/classes/class.ilForumTopic.php [hunk 7]
```diff
@@ -997,7 +1041,7 @@ public function updateThreadTitle() : void
             ['thr_pk' => ['integer', $this->getId()]]
         );
 
-        $first_node = $this->getFirstPostNode();
+        $first_node = $this->getPostRootNode();
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumTopic.php]:
```suggestion
        $first_node = $this->getFirstVisiblePostNode();
```

**Review by mjansenDatabay — APPROVED** [review]
Thx @fhelfer

**Pull request merged**